### PR TITLE
Enable feature map/saliency map output for OpenVINO inference task

### PR DESCRIPTION
**Pull request by eugene123tw**
Enable feature map/saliency map output for OpenVINO/PyTorch inference task

### CODE REVIEW

**Comment by eugene123tw** [comment]
run ote-test

File: external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    def _get_outputs(self) -> Dict:
+        """Match the output names with graph node index"""
+        output_match_dict = {}
+        output_names = ["boxes", "labels", "feature_vector", "saliency_map"]
+        for output_name in output_names:
+            for node_name, node_meta in self.outputs.items():
+                if output_name in node_meta.names:
+                    output_match_dict[output_name] = node_name
+                    break
+        return output_match_dict
```
Comment (goodsong81) [inline external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py:58]: I don't know much about the model APIs. Don't we need this for ATSS & YOLOX?

File: external/mmdetection/detection_tasks/apis/detection/inference_task.py [hunk 2]
```diff
@@ -306,25 +319,35 @@ def _infer_detector(model: torch.nn.Module, config: Config, dataset: DatasetEnti
...
                 assert feature_vector.size(0) == 1
             feature_vectors.append(feature_vector.view(-1).detach().cpu().numpy())
 
         def dummy_dump_features_hook(mod, inp, out):
             feature_vectors.append(None)
 
-        hook = dump_features_hook if dump_features else dummy_dump_features_hook
+        def dump_saliency_hook(mod, inp, out):
+            with torch.no_grad():
+                saliency_map = get_saliency_map(out[0])
```
Comment (goodsong81) [inline external/mmdetection/detection_tasks/apis/detection/inference_task.py]: Same question: out[0] or out[-1]?
Or am I misunderstanding the index? Is it layer index?

File: external/mmdetection/detection_tasks/apis/detection/ote_utils.py [hunk 3]
```diff
@@ -174,3 +175,39 @@ def on_train_end(self, logs=None):
...
+
+    Returns:
+        torch.Tensor: feature vector(representation vector)
+    """
+    pooled_features = [
+        torch.nn.functional.adaptive_avg_pool2d(feat_map, (1, 1))
+        for feat_map in feature_maps
+    ]
+    feature_vector = torch.cat(pooled_features, dim=1)
+    return feature_vector
```
Comment (goodsong81) [inline external/mmdetection/detection_tasks/apis/detection/ote_utils.py]: I saw these functions in mmdet PRs, too. Do we need to keep them both?

File: external/mmdetection/detection_tasks/apis/detection/inference_task.py (hunk 2 above)
Comment (eugene123tw) [reply external/mmdetection/detection_tasks/apis/detection/inference_task.py]: Docstring added. Please also check my reply in mmdetecion PR (https://github.com/openvinotoolkit/mmdetection/pull/352), we can discuss it there. I'm currently taking the largest feature map at index 0 i.e. `out[0]`.

File: external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py (hunk 1 above)
Comment (eugene123tw) [reply external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py:58]: ATSS model is based on SSD, so no need for ATSS. That said, we need to implement a YOLOX Model API wrapper as it is using SSD ModelAPI and it's not correct.

File: external/mmdetection/detection_tasks/apis/detection/ote_utils.py (hunk 3 above)
Comment (eugene123tw) [reply external/mmdetection/detection_tasks/apis/detection/ote_utils.py]: Fair point and removed

**Comment by eugene123tw** [comment]
run ote-test

File: external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py (hunk 1 above)
Comment (goodsong81) [reply external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py:58]: OK. Please update the wrapper for YOLOX, too. Thanks!

File: external/mmdetection/detection_tasks/apis/detection/inference_task.py (hunk 2 above)
Comment (goodsong81) [reply external/mmdetection/detection_tasks/apis/detection/inference_task.py]: As I commented in the mmdet PR, let's use out[-1] which has most high-level semantic.

Comment (eugene123tw) [reply external/mmdetection/detection_tasks/apis/detection/inference_task.py]: Updated

**Comment by eugene123tw** [comment]
run ote-test

File: external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py (hunk 1 above)
Comment (eugene123tw) [reply external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py:58]: The changes in YOLOX will mostly be in the exportable code than the saliency map code. Could we address this in a different PR?

Comment (goodsong81) [reply external/mmdetection/detection_tasks/apis/detection/model_wrappers/openvino_models.py:58]: I think it would be OK.

**Review by goodsong81 — APPROVED** [review]
LGTM!

**Comment by harimkang** [comment]
Currently, there are many similar parts to OTE Detection's code in the detection interference part of MPA. There seems to be no problem with this PR itself, but I think it can affect MPA due to the modification of mmdetection, so should we look at it after the merge? Or do you think there's anything that can affect? (Currently, the test of MPA is not conducted in CI test, so it is difficult to check..) /// Of course, it seems okay to fix the code on the MPA side in another PR.

If possible, it would be a good idea to run and check the detection sample of MPA
```
CUDA_VISIBLE_DEVICES=0 python external/model-preparation-algorithm/mpa_tasks/samples/det_cls_il_sample.py external/model-preparation-algorithm/configs/detection/mobilenetv2_atss_cls_incr/template_experimental.yaml
```

**Review by harimkang — APPROVED** [review]
LGTM.
There seems to be no issue when I run the task sample code on the MPA side based on (the PR branch + latest develop rebase).

**Comment by JihwanEom** [comment]
I merged it by mistake, but it's all approved, so I'll leave it.

**Comment by eugene123tw** [comment]
@goodsong81 The branch in mmdetection submodule is not rebased to `ote`. Will this cause a problem?

**Comment by goodsong81** [comment]
> @goodsong81 The branch in mmdetection submodule is not rebased to `ote`. Will this cause a problem?

I would let @JihwanEom update the mmdetation commit in other PR. Thank!

**Comment by JihwanEom** [comment]
I'll carefully check this. Sorry for inconvenience.

**Pull request merged**